### PR TITLE
System.IO.Pipes: Add additional Create overloads for Mono

### PR DESCRIPTION
**Pull request by filipnavara**
Rationale: There are some missing constructor overloads in the pipe classes that deal with `PipeSecurity` (#30170).

`PipeSecurity` is inherited from `NativeObjectSecurity`, which throws `PlatformNotSupportedException` in `InternalSet` on anything but Windows. Since the implementation doesn't really seem cross-platform it is questionable whether the overloads would make sense on CoreFX.

However, Mono implements the NetFX API surface and thus it has to provide implementation for the constructors. It seems reasonable to limit such implementation to Windows and throw `PlatformNotSupportedException` on anything else. In order to provide the Windows implementation some code must be exposed in CoreFX that would be used in Mono to implement the remaining overloads.

### CODE REVIEW

**Comment by filipnavara** [comment]
@dotnet-bot test Packaging All Configurations x64 Debug Build

File: src/System.IO.Pipes/src/System/IO/Pipes/NamedPipeServerStream.Windows.cs [hunk 1]
```diff
@@ -39,9 +47,7 @@ public sealed partial class NamedPipeServerStream : PipeStream
                 throw new ArgumentOutOfRangeException(nameof(pipeName), SR.ArgumentOutOfRange_AnonymousReserved);
             }
 
-            PipeSecurity pipeSecurity = null;
-
-            if (IsCurrentUserOnly)
+            if (IsCurrentUserOnly && pipeSecurity == null)
```
Comment (pjanotti) [inline src/System.IO.Pipes/src/System/IO/Pipes/NamedPipeServerStream.Windows.cs]: In CoreFx an assert will be good to ensure that whenever `IsCurrentUserOnly` is true `pipeSecurity` is also null. In Mono what will be the behavior for the constructors that the respective parameters?

File: src/System.IO.Pipes/src/System/IO/Pipes/AnonymousPipeServerStream.Windows.cs [hunk 2]
```diff
@@ -16,6 +17,12 @@ public sealed partial class AnonymousPipeServerStream : PipeStream
     {
         // Creates the anonymous pipe.
         private void Create(PipeDirection direction, HandleInheritability inheritability, int bufferSize)
+        {
+            Create(direction, inheritability, bufferSize, null);
+        }
+
+        // Creates the anonymous pipe.
+        private void Create(PipeDirection direction, HandleInheritability inheritability, int bufferSize, PipeSecurity pipeSecurity)
```
Comment (pjanotti) [inline src/System.IO.Pipes/src/System/IO/Pipes/AnonymousPipeServerStream.Windows.cs]: Add comment for the parameter `pipeSecurity` to indicate that it is used by public constructors in Mono.

File: src/System.IO.Pipes/src/System/IO/Pipes/NamedPipeServerStream.Windows.cs (hunk 1 above)
Comment (filipnavara) [reply src/System.IO.Pipes/src/System/IO/Pipes/NamedPipeServerStream.Windows.cs]: The legacy constructors won't allow `CurrentUserOnly` mode if `pipeSecurity != null`.

Comment (pjanotti) [reply src/System.IO.Pipes/src/System/IO/Pipes/NamedPipeServerStream.Windows.cs]: Could you please add a Debug.Assert - other than that LGTM.

Comment (filipnavara) [reply src/System.IO.Pipes/src/System/IO/Pipes/NamedPipeServerStream.Windows.cs]: Added the assert here: https://github.com/dotnet/corefx/pull/31118/files#diff-a7e95545fe2b63c35aaee6c4c3de3cffR53

Comment (pjanotti) [reply src/System.IO.Pipes/src/System/IO/Pipes/NamedPipeServerStream.Windows.cs]: Ops, missed that.

**Comment by filipnavara** [comment]
Thanks for review!

**Comment by pjanotti** [comment]
@dotnet-bot test Packaging All Configurations x64 Debug Build

**Pull request merged**